// wecom/dist/modules/oa/checkin/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Checkin = void 0;
var wecom_1 = require("../../../wecom");
var Checkin = /** @class */ (function (_super) {
    __extends(Checkin, _super);
    function Checkin(config) {
        return _super.call(this, config) || this;
    }
    /**
     * @description 获取打卡数据
     * @param {QueryCheckinData} data
     * @return {*}  {Promise<AxiosResponse<ICheckinDataRet>>}
     * @memberof Checkin
     */
    Checkin.prototype.getCheckinData = function (data) {
        return this.request({
            url: "checkin/getcheckindata",
            method: "POST",
            data: data,
        });
    };
    /**
     * @description 获取打卡规则
     * @param {QueryCheckinOption} data
     * @return {*}  {Promise<AxiosResponse<ICheckinOptionRet>>}
     * @memberof Checkin
     */
    Checkin.prototype.getCheckinOption = function (data) {
        return this.request({
            url: "checkin/getcheckinoption",
            method: "POST",
            data: data,
        });
    };
    return Checkin;
}(wecom_1.Wecom));
exports.Checkin = Checkin;
